contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1727,6 +1730,204 @@ object_name(PG_FUNCTION_ARGS)
+        result = (*common_utility_plugin_ptr->get_tsql_datatype_oid) (object_name);
+
+        // If null result, then it is not system datatype and now search with default schema in pg_type
+        if (!OidIsValid(result))
+        {
+            /* find the default schema for current user and get physical schema name */
+            const char *user = get_user_for_database(db_name);
+            const char *guest_role_name = get_guest_role_name(db_name);
+
+            if (!user)
+            {
+                pfree(db_name);
+                pfree(schema_name);
+                pfree(object_name);
+                PG_RETURN_NULL();
+            }
+            else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+            {
+                physical_schema_name = pstrdup(get_guest_schema_name(db_name));
+            }
+            else
+            {
+                pfree(schema_name);
+                schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+                physical_schema_name = get_physical_schema_name(db_name, schema_name);
+            }
+        }
+        else
+        {
+            pfree(db_name);
+            pfree(schema_name);
+            pfree(object_name);
+            PG_RETURN_INT32(result);
+        }
+    }
+    else
+    {
+        physical_schema_name = get_physical_schema_name(db_name, schema_name);
+    }
+
+    /* get schema oid from physical schema name, it will return InvalidOid if user don't have lookup access */
+    schema_oid = get_namespace_oid(physical_schema_name, true);
+    pfree(db_name);
+    pfree(physical_schema_name);
+
+    if (!OidIsValid(schema_oid) || pg_namespace_aclcheck(schema_oid, user_id, ACL_USAGE) != ACLCHECK_OK)
+    {
+        pfree(schema_name);
+        pfree(object_name);
+        PG_RETURN_NULL();
+    }
+    // If schema is 'sys' or 'pg_catalog' then search in typecode list, else in pg_type.
+    if(!strcmp(schema_name, "sys") || !strcmp(schema_name, "pg_catalog"))
+    {
+        result = (*common_utility_plugin_ptr->get_tsql_datatype_oid) (object_name);
+    }
+    else
+    {
+        result = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid,
+                                    CStringGetDatum(object_name), ObjectIdGetDatum(schema_oid));
+    }
+    pfree(schema_name);
+    pfree(object_name);
+    if (OidIsValid(result) && pg_type_aclcheck(result, user_id, ACL_USAGE) == ACLCHECK_OK)
+        PG_RETURN_INT32(result);
+    else
+        PG_RETURN_NULL();
+}
+
+/*
+ * type_name
+ *      returns the type name with type id as input
+ * Returns NULL
+ *      if there is no such type
+ *      if user don't have right permission
+ */
+Datum
+type_name(PG_FUNCTION_ARGS)
+{
+    Datum       type_id = PG_GETARG_DATUM(0);
+    Datum       tsql_typename;
+    HeapTuple   tuple;
+    Oid         user_id = GetUserId();
+    char        *result = NULL;
+
+    LOCAL_FCINFO(fcinfo1, 1);
+
+    if (type_id < 0)
```

Comment:
Let us take oid as input datatype instead of int. I think then this check would be redundant. After this change, what is the behaviour for negative inputs?